File: 94BinaryTreeInorderTraversal.cpp
#include <iostream>
#include <set>
#include <vector>
#include <iterator>
#include <utility>
#include <cstdlib>
#include <limits>
#include <functional>
#include <stack>
#include <unordered_set>

using namespace std;

template<class Container>
void printContainer(const Container& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[";
        }
        cout << *it;
        if (it == prev(container.end())) {
            cout << "]" << endl;
        } else {
            cout << ", ";
        }
    }
}

/**
 * Definition for a binary tree node.
 */
 struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode() : val(0), left(nullptr), right(nullptr) {}
     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> output;

        /*
         * recursive solution
        function<void(TreeNode*)> traverse = [&output, &traverse](TreeNode* node) {
            if (node == nullptr) {
                return;
            }

            traverse(node->left);
            output.push_back(node->val);
            traverse(node->right);
        };

        traverse(root);
         */

        if (root == nullptr) {
            return output;
        }

        //iterative solution
        stack<TreeNode*> deferedNodes;
        unordered_set<TreeNode*> visitedNodes;

        while (true) {
            if (root->left != nullptr &&
                    visitedNodes.find(root->left) == visitedNodes.end()) {
                deferedNodes.push(root);
                visitedNodes.insert(root->left);
                root = root->left;
            } else {
                output.push_back(root->val);
                if (root->right != nullptr) {
                    root = root->right;
                } else {
                    if (deferedNodes.empty()) {
                        break;
                    } else {
                        root = deferedNodes.top();
                        deferedNodes.pop();
                    }
                }
            }
        }


        return output;
    }
};


int main(){
    Solution s;
    TreeNode t1(3);
    TreeNode t2(2, &t1, nullptr);
    TreeNode root(1, nullptr, &t2);
    auto output = s.inorderTraversal(&root);

    auto output1 = s.inorderTraversal(nullptr);

    TreeNode root2(1);
    auto output2 = s.inorderTraversal(&root2);

    TreeNode t3(2);
    TreeNode root3(1, &t3, nullptr);
    auto output3 = s.inorderTraversal(&root3);


    TreeNode t4(2);
    TreeNode root4(1, nullptr, &t4);
    auto output4 = s.inorderTraversal(&root4);

    TreeNode t5(1);
    TreeNode t6(2);
    TreeNode root5(3, &t5, &t6);
    auto output5 = s.inorderTraversal(&root5);

    cout << "*************output*************" << endl;
    printContainer(output);
    printContainer(output1);
    printContainer(output2);
    printContainer(output3);
    printContainer(output4);
    printContainer(output5);

    return 0;
}

File: go/94BinaryTreeInorderTraversal.go
package main

import (
	"container/list"
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func inorderTraversal(root *TreeNode) []int {
	output := []int{}

	/*
			//recursive solution
			var traverse func(*TreeNode, *[]int)
			traverse = func(node *TreeNode, nodeVal *[]int) {
				if node == nil {
					return
				}

				traverse(node.Left, nodeVal)
				*nodeVal = append(*nodeVal, node.Val)
				traverse(node.Right, nodeVal)
			}

		traverse(root, &output)
	*/

	//iterative solution
	staPendingParentNodes := list.New()
	mapReachedNode := make(map[*TreeNode]bool)
	curNode := root

	for curNode != nil {
		leftNodeIsEmptyOrReached := false
		if curNode.Left == nil {
			leftNodeIsEmptyOrReached = true
		}
		if _, ok := mapReachedNode[curNode.Left]; ok {
			leftNodeIsEmptyOrReached = true
		}

		if leftNodeIsEmptyOrReached {
			output = append(output, curNode.Val)
			mapReachedNode[curNode] = true

			if curNode.Right != nil {
				curNode = curNode.Right
			} else {
				if staPendingParentNodes.Front() != nil {
					curNode = staPendingParentNodes.Front().Value.(*TreeNode)
					staPendingParentNodes.Remove(staPendingParentNodes.Front())
				} else { // the end of traversal
					curNode = nil
				}
			}
		} else {
			staPendingParentNodes.PushFront(curNode)
			curNode = curNode.Left
		}
	}

	return output
}

func main() {
	output := inorderTraversal(
		&TreeNode{
			Val:  1,
			Left: nil,
			Right: &TreeNode{
				Val: 2,
				Left: &TreeNode{
					Val:   3,
					Left:  nil,
					Right: nil,
				},
				Right: nil,
			},
		})
	output1 := inorderTraversal(nil)
	output2 := inorderTraversal(
		&TreeNode{
			Val:   1,
			Left:  nil,
			Right: nil,
		})
	output3 := inorderTraversal(
		&TreeNode{
			Val: 1,
			Left: &TreeNode{
				Val:   2,
				Left:  nil,
				Right: nil,
			},
			Right: nil,
		})
	output4 := inorderTraversal(
		&TreeNode{
			Val:  1,
			Left: nil,
			Right: &TreeNode{
				Val:   2,
				Left:  nil,
				Right: nil,
			},
		})
	output5 := inorderTraversal(
		&TreeNode{
			Val: 3,
			Left: &TreeNode{
				Val:   1,
				Left:  nil,
				Right: nil,
			},
			Right: &TreeNode{
				Val:   2,
				Left:  nil,
				Right: nil,
			},
		})

	fmt.Println("\n******** output *********\n", output, output1, output2, output3, output4, output5)
}
